File: src/frontend/lib/data/models/save.dart
import 'package:json_annotation/json_annotation.dart';

part 'save.g.dart';

@JsonSerializable()
class Save {
  final String saveId;
  final String playerId;
  final String saveName;
  final String gameTime;
  final int playerLevel;
  final Map<String, dynamic> attributes;
  final String currentChapter;
  final List<dynamic> inventory;
  final List<dynamic> quests;
  final List<dynamic> achievements;
  final int playTime;
  final DateTime createdAt;
  final DateTime updatedAt;

  Save({
    required this.saveId,
    required this.playerId,
    required this.saveName,
    required this.gameTime,
    required this.playerLevel,
    required this.attributes,
    required this.currentChapter,
    required this.inventory,
    required this.quests,
    required this.achievements,
    required this.playTime,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Save.fromJson(Map<String, dynamic> json) => _$SaveFromJson(json);

  Map<String, dynamic> toJson() => _$SaveToJson(this);

  Save copyWith({
    String? saveId,
    String? playerId,
    String? saveName,
    String? gameTime,
    int? playerLevel,
    Map<String, dynamic>? attributes,
    String? currentChapter,
    List<dynamic>? inventory,
    List<dynamic>? quests,
    List<dynamic>? achievements,
    int? playTime,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Save(
      saveId: saveId ?? this.saveId,
      playerId: playerId ?? this.playerId,
      saveName: saveName ?? this.saveName,
      gameTime: gameTime ?? this.gameTime,
      playerLevel: playerLevel ?? this.playerLevel,
      attributes: attributes ?? this.attributes,
      currentChapter: currentChapter ?? this.currentChapter,
      inventory: inventory ?? this.inventory,
      quests: quests ?? this.quests,
      achievements: achievements ?? this.achievements,
      playTime: playTime ?? this.playTime,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: src/backend/src/models/Save.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface ISave extends Document {
  saveId: string;
  playerId: string;
  saveName: string;
  gameTime: string;
  playerLevel: number;
  attributes: {
    chakra: number;
    ninjutsu: number;
    taijutsu: number;
    intelligence: number;
    speed: number;
    luck: number;
  };
  currentChapter: string;
  inventory: any[];
  quests: any[];
  achievements: any[];
  playTime: number;
  isCloud: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const SaveSchema: Schema = new Schema(
  {
    saveId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    playerId: {
      type: String,
      required: true,
      index: true,
    },
    saveName: {
      type: String,
      required: true,
      maxlength: 50,
    },
    gameTime: {
      type: String,
      required: true,
    },
    playerLevel: {
      type: Number,
      default: 1,
    },
    attributes: {
      chakra: { type: Number, default: 100 },
      ninjutsu: { type: Number, default: 50 },
      taijutsu: { type: Number, default: 50 },
      intelligence: { type: Number, default: 50 },
      speed: { type: Number, default: 50 },
      luck: { type: Number, default: 50 },
    },
    currentChapter: {
      type: String,
      required: true,
    },
    inventory: {
      type: [Schema.Types.Mixed],
      default: [],
    },
    quests: {
      type: [Schema.Types.Mixed],
      default: [],
    },
    achievements: {
      type: [Schema.Types.Mixed],
      default: [],
    },
    playTime: {
      type: Number,
      default: 0,
    },
    isCloud: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  },
);

// 创建索引
SaveSchema.index({ playerId: 1, updatedAt: -1 });
SaveSchema.index({ isCloud: 1 });

export default mongoose.model<ISave>('Save', SaveSchema);
